switching dependencies over to those used elsewhere at UKAEA

smason opened the pull request:
basically drop everything I can that isn't in the STL, use CMake for building, and catch2 for unit tests

smason committed:
much fiddling, unit tests still to do

catch2 doesn't seem nearly as nice as doctest, might be quite a bit
more work

smason committed:
switch unit tests from doctest to catch2

smason committed:
explicitly install g++

not sure how this was getting picked up before, maybe via python?

makeclean commented on CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+include_directories(SYSTEM "include")
+
+find_package(Catch2 REQUIRED)
+add_executable(test_runner src/test_runner.cpp src/document.cpp src/utils.cpp)
```
Might be worth having this main cmakefile call a subsequent cmakefile in the src directory which is closer to the tools?

makeclean replied:
Lets also submodule catch2 for simplicity

smason replied:
the file is only ~30 lines long, not sure if it's worth splitting up is it?  if it was hundreds of lines I could see an argument, but I'm not sure how another small file would improve clarity

makeclean replied:
more for stylistic and good practise reasons, best to have a root level cmake file, which calls those from higher directories

makeclean commented on include/cxx_argp_parser.h:
```diff
@@ -0,0 +1,246 @@
+//
+// Copyright (C) 2018-2019 Patrick Boettcher <p@yai.se>
+//
+// SPDX-License-Identifier: LGPL-3.0
```
This is LGPL-3, this isn't compatible with the main licence is it?

smason replied:
oops, didn't notice the v3 part, I'll make this a submodule then

makeclean commented on src/document.cpp:
```diff
@@ -1,13 +1,9 @@
-#ifdef INCLUDE_DOCTESTS
-#include <doctest/doctest.h>
+#ifdef INCLUDE_TESTS
+#include <catch2/catch.hpp>
```
why is this file called document?

smason replied:
early versions of the code just had the `document` class in, but have added other functionality and not renamed the file.  not sure maybe `geometry.cpp`?

makeclean commented on src/imprint_solids.cpp:
```diff
@@ -84,19 +86,26 @@ imprint(document &doc)
-			->required()
-			->option_text("output.brep");
-		CLI11_PARSE(app, argc, argv);
+		const char *doc = (
```
Does it really have to be char*?

smason replied:
what else should it be?  `static const char[]`?

makeclean replied:
i was thinking std::string

smason replied:
but `cxx_argp::parse` takes a `const char*` as a parameter, so that just seems to be making unnecessary work for the compiler to undo

makeclean replied:
fair enough, I try to avoid char* as much as possible, but thats potentially a religious position :)

makeclean commented on src/overlap_checker.cpp:
```diff
@@ -177,81 +153,98 @@ are_bboxs_disjoint(const Bnd_OBB &b1, const Bnd_OBB& b2, double tolerance)
 
 	std::string path_in;
-	unsigned num_parallel_jobs;
+	unsigned num_parallel_jobs = 4;
```
we should probably assume 1 if not set?

smason replied:
will do, might be a safer default!

makeclean commented on src/overlap_checker.cpp:
```diff
@@ -177,81 +153,98 @@ are_bboxs_disjoint(const Bnd_OBB &b1, const Bnd_OBB& b2, double tolerance)
+		assert(args.size() == 1);
+		path_in = args[0];
+
+		if (num_parallel_jobs < 1 || num_parallel_jobs > 1024) {
```
why 1024?

smason replied:
sanity check mostly, not sure if there are many systems with close to that many cores available.

makeclean replied:
Is it just threads?

makeclean replied:
smells like a magic number, is there no compiler macro to check how many cpus are the on the machine?

smason replied:
yes it's a magic number, so should probably be in a macro.  it's just there to stopping the user from shooting themselves in the foot

I look at `max_parallel` later to generate a more appropriate warning later on

makeclean commented on src/overlap_checker.cpp:
```diff
@@ -177,81 +153,98 @@ are_bboxs_disjoint(const Bnd_OBB &b1, const Bnd_OBB& b2, double tolerance)
-				"requesting significantly more than the number of cores ({} > {}) is unlikely to help",
-				num_parallel_jobs, max_parallel);
+		if (max_parallel && num_parallel_jobs > max_parallel * 2 + 4) {
+			LOG(WARNING)
```
why?

smason replied:
it's CPU bound code.  I didn't want to warn users if they picked a couple more threads (so the scheduler always has work to do) and wasn't sure if/how hyperthreading was going to picked up

makeclean replied:
are you threading with OpenMP?

smason replied:
just `pthreads` as it didn't seem complicated enough to warrant anything else

makeclean replied:
ok, the reason I ask is openmp has a function that returns the number of cores, I guess pthreads has the same? SO suggests https://stackoverflow.com/questions/150355/programmatically-find-the-number-of-cores-on-a-machine

smason replied:
yes, `max_parallel` is set from that standard `std::thread::hardware_concurrency` API.  note that it reports hyper-threads instead of just "cores" for me under Linux

I wanted to let the user specify larger values for the number of parallel jobs (which is especially useful for my testing) but thought it would be useful to fail when they start requesting silly big numbers

makeclean reviewed (commented):
Looks ok, going to build locally, there are few odd things that Im sure aren't odd when I've got some context :)

makeclean commented:
Would it be worth registering these tests to ctest?

makeclean commented:
For the various tools, lets alias '-h' to '--help'

makeclean commented:
Text for overlap_checker help 
```
./overlap_checker --help
Usage: overlap_checker [OPTION...] input.brep
Perform imprinting of BREP shapes.

      --bbox-clearance=C     Bounding-boxes closer than C[=0.5] will be checked
                             for overlaps
      --imprint-tolerance=T  Faces, edges, and verticies will be merged when
                             closer than T[=0.001]
  -j, --jobs=N               Parallelise over N[=4] threads
      --max-common-volume-ratio=R
                             Imprinted volume with ratio <R[=0.01] is
                             considered acceptable
  -?, --help                 Give this help list
      --usage                Give a short usage message

Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
```
Says that its for imprinting stuff, which it isn't

makeclean commented:
overlap_checker - default seems to be debug being on, should this not be off by default?

makeclean commented:
No useful summary from overlap checker, I would expect something like a list of overlaps, or a message to say there weren't any

smason commented:
> No useful summary from overlap checker

what would a useful summary look like?  it did follow the old school unix paradigm of no output indicates success, at least it did until I broke the code that sets logging levels

it'll [output a line](https://github.com/ukaea/overlap_checker/blob/8b6ce6c6ee3319b58da5339405fe483bde7197fd/src/overlap_checker.cpp#L348-L351) if there are problems, are you saying it should always output a line, listing things like number of comparisons done, number of shapes touching, with small overlaps, etc?

re. the other comments, I'll see what I can do this afternoon

smason committed:
use submodules for aixlog and cxx_argp

think I needed to change cxx_argp before, so pulled in directly, but
now changes have been merged upstream

smason committed:
forgot to tell CI to clone recursively

smason committed:
use github rather than gitlab syntax

smason committed:
don't specify a container

smason committed:
sudo needed to run apt-get?

smason committed:
catch2 is called catch in ubuntu!

smason committed:
switch to Catch2 as a git submodule

smason committed:
another way to checkout submodules

smason committed:
sudo not needed in a container

smason committed:
is this the magic CI incanation

smason committed:
apparently runs-on is always needed, even with a container

smason committed:
oops, was compiling with my system catch2 not the submodule

also switch to the new v3 catch2 headers

smason committed:
hide the const char* behind a const auto!

smason committed:
safer default on the number of parallel jobs

smason committed:
more explicit that the 1024 limit is a magic number

smason committed:
removing reference to upstream cxx_argp

smason committed:
use my fork of cxx_argp to provide more standard help flag

smason committed:
udpate to latest cxx_argp

makeclean commented:
I guess I would expect a summary table as a default option, with some useful stats, like number of overlaps, number of overlap checks, number of coincidences and so on?

smason committed:
split up CMakeLists.txt files

smason committed:
improve help of overlap_checker

smason committed:
missed '\n' in logging

smason committed:
log a summary line of work done

smason commented:
> No useful summary from overlap checker, I would expect something like a list of overlaps, or a message to say there weren't any

you could always `tee` the CSV output to watch progress

that said, I've added a summary line in 074b8c1ae540efce1940227fd77dee15097a5feb that outputs:

> 202.961 [Info] processing summary: bbox tests=525825, intersection tests=3404, touching=158, overlapping=1, bad overlaps=0, tests failed=0

smason committed:
rename document.[ch]pp to geometry

smason commented:
@makeclean think that's all done now

please let me know if I've missed anything!

smason committed:
swicth to ctest

not sure why I didn't need NAME and COMMAND in there before

makeclean commented:
The end of run summary is great!

Im seeing a bunch of things like
```
204.528 [Warn]   435-129 imprint failed with (73 filler and 0 common) warnings
204.528 [Error]   435-129 failed to classify overlap
```
What is a filler warning?

I'm not sure if you want to branch some of the following of here into issues, or just fix em all here?

* the overlap check should read a STEP file in addition to a BREP (probably more importantly a STEP file)
* the imprint tool should read a STEP file in addition to a BREP (probably more importantly a STEP file)
* im getting output to screen a stream of output to screen like <x,y,touch> I presume this should be going elsewhere?

I'm noticing a tool step_to_brep - I presume I should run this on STEP files first, would it not be easier (better?) to have filename based dispactch in the overlap and imprint tools to either convert a step file to brep then run?

makeclean commented:
Hmm, I actually quite like the the one tool one job, unix pipeline maybe type dispatch would be overkill. Though some additonal verbositry form the input tool would be good, I've currently no idea what its doing, or how long it might take?

smason commented:
> What is a filler warning?

it's coming from the deep guts of opencascade, I'm reporting it because it was easy to pull out and in the hope it'd be useful for debugging if/when something like this fails

maybe open another issue with a pointer to the file an I can have a look?

smason commented:
the main reason for separating things out is to help with materials later, I've added some detail as part of another merge: https://github.com/ukaea/overlap_checker/tree/shape-merge-tool#file-formats
 
i.e. `step_to_brep` flattens the file into a list of shapes, then all other tools keep shapes the "same", when I start handling materials it'll hopefully be an easy job to match things back up afterwards
